zerver/tests/test_email_notifications.py

```diff
@@ -234,6 +236,198 @@ def test_send_custom_email_dry_run(self) -> None:
+                event_type=RealmAuditLog.CUSTOM_EMAIL_SENT,
+            )
+            self.assert_length(logged_events, 2)
+            user_id_set = {logged_event.acting_user_id for logged_event in logged_events}
+            expected_user_id_set = {hamlet.id, zoe.id}
+            self.assertEqual(user_id_set, expected_user_id_set)
+            hash_code_set = {
+                logged_event.extra_data.get("email_template_digest")
+                for logged_event in logged_events
+            }
+            self.assertEqual(hash_code_set, {hash_code})
+            logged_events = RealmAuditLog.objects.filter(
+                realm=mit_realm,
+                event_type=RealmAuditLog.CUSTOM_EMAIL_SENT,
+            )
+            self.assert_length(logged_events, 2)
+            user_id_set = {logged_event.acting_user_id for logged_event in logged_events}
+            expected_user_id_set = {hamlet_mit.id, zoe_mit.id}
+            self.assertEqual(user_id_set, expected_user_id_set)
+            hash_code_set = {
+                logged_event.extra_data.get("email_template_digest")
+                for logged_event in logged_events
+            }
+            self.assertEqual(hash_code_set, {hash_code})
+
+            # sending the same email to 5 users, only aaron is new.
+            aaron = self.example_user("aaron")
+            email_count += 1
+            TestCustomEmails.send_email_helper(
+                users=[hamlet, hamlet_mit, aaron, zoe, zoe_mit],
+                template_file=markdown_template.name,
+                distinct_email=True,
+            )
+            self.assert_length(mail.outbox, email_count)
+            logged_events = RealmAuditLog.objects.filter(
+                realm=example_realm,
+                event_type=RealmAuditLog.CUSTOM_EMAIL_SENT,
+            )
+            self.assert_length(logged_events, 3)
+            user_id_set = {logged_event.acting_user_id for logged_event in logged_events}
+            expected_user_id_set = {hamlet.id, zoe.id, aaron.id}
+            self.assertEqual(user_id_set, expected_user_id_set)
+            hash_code_set = {
+                logged_event.extra_data.get("email_template_digest")
+                for logged_event in logged_events
+            }
+            self.assertEqual(hash_code_set, {hash_code})
+
+            logged_events = RealmAuditLog.objects.filter(
+                realm=mit_realm,
+                event_type=RealmAuditLog.CUSTOM_EMAIL_SENT,
+            )
+            self.assert_length(logged_events, 2)
+            user_id_set = {logged_event.acting_user_id for logged_event in logged_events}
+            expected_user_id_set = {hamlet_mit.id, zoe_mit.id}
+            self.assertEqual(user_id_set, expected_user_id_set)
+            hash_code_set = {
+                logged_event.extra_data.get("email_template_digest")
+                for logged_event in logged_events
+            }
+            self.assertEqual(hash_code_set, {hash_code})
```

Comment:
I think these could be better done as end-to-end tests that mainly verify that duplicate emails aren't sent, with checking the state of the audit logs being a side detail. See https://zulip.readthedocs.io/en/latest/testing/philosophy.html

Reply:
@timabbott Can you clarify? In tests, I checked the duplicate emails weren't sent  by : 
assert_length(mail.outbox, email_count) with email_count tracks the expected emails sent and mail.outbox tracks the actual emails sent.